File: backend/app/services/file_types.py
from __future__ import annotations

from pathlib import Path
from typing import Optional


STRUCTURE_EXTENSIONS = {
    ".xyz": "xyz",
    ".pdb": "pdb",
    ".mol": "mol",
    ".sdf": "sdf",
    ".cif": "cif",
    ".traj": "traj",
    ".extxyz": "extxyz",
    ".vasp": "vasp",
    ".xsf": "xsf",
    ".cube": "cube",
    ".gen": "gen",
    ".db": "db",
    ".lammps": "lammps",
    ".dump": "lammps-dump-text",
    ".xml": "xml",
    ".gjf": "gaussian-in",
    ".com": "gaussian-in",
    ".fdf": "fdf",
    ".pwi": "espresso-in",
}

TEXT_EXTENSIONS = {
    ".txt",
    ".log",
    ".out",
    ".in",
    ".inp",
    ".sh",
    ".py",
    ".json",
    ".yaml",
    ".yml",
    ".md",
    ".dat",
    ".csv",
    ".err",
}

TEXT_NAMES = {
    "INCAR",
    "KPOINTS",
    "OSZICAR",
    "POTCAR",
}

ASE_STRUCTURE_NAMES = {
    "OUTCAR",
    "XDATCAR",
    "vasp.xml",
}


def is_forced_structure_name(path: Path) -> bool:
    name = path.name.upper()
    return "POSCAR" in name or "CONTCAR" in name


def detect_preview(path: Path, *, is_dir: bool = False) -> tuple[str, Optional[str]]:
    if is_dir:
        return "directory", None

    if is_forced_structure_name(path):
        return "structure", None

    if path.name.upper() in ASE_STRUCTURE_NAMES or path.name.lower() in ASE_STRUCTURE_NAMES:
        return "structure", None

    suffix = path.suffix.lower()
    if suffix in STRUCTURE_EXTENSIONS:
        return "structure", STRUCTURE_EXTENSIONS[suffix]
    if suffix in TEXT_EXTENSIONS or path.name.upper() in TEXT_NAMES:
        return "text", None
    return "file", None

File: frontend/src/api/fileTypes.ts
import type { FileItem, PreviewType } from './files'

const STRUCTURE_FALLBACK_EXTENSIONS = new Set([
  '.xyz',
  '.pdb',
  '.mol',
  '.sdf',
  '.cif',
  '.traj',
  '.extxyz',
  '.vasp',
  '.xsf',
  '.xsd',
  '.cube',
  '.gen',
  '.db',
  '.lammps',
  '.dump',
  '.xml',
  '.gjf',
  '.com',
  '.fdf',
  '.pwi'
])

const STRUCTURE_FALLBACK_NAMES = new Set(['xdatcar', 'outcar', 'vasp.xml'])

export function pathBaseName(path: string) {
  return path.split(/[\\/]/).pop() ?? path
}

export function extensionFromName(name: string) {
  const index = name.lastIndexOf('.')
  return index >= 0 ? name.slice(index).toLowerCase() : ''
}

export function isForcedStructureName(name: string) {
  const normalized = name.toUpperCase()
  return normalized.includes('POSCAR') || normalized.includes('CONTCAR')
}

export function isStructurePreviewItem(item: Pick<FileItem, 'name' | 'preview_type'>) {
  return item.preview_type === 'structure' || isForcedStructureName(item.name)
}

export function isStructurePreviewPath(path: string, previewType?: PreviewType | null) {
  if (previewType === 'structure') return true
  const name = pathBaseName(path).toLowerCase()
  return (
    isForcedStructureName(name) ||
    STRUCTURE_FALLBACK_NAMES.has(name) ||
    STRUCTURE_FALLBACK_EXTENSIONS.has(extensionFromName(name))
  )
}
